druid-shell/src/keyboard.rs

```diff
@@ -0,0 +1,220 @@
+// Copyright 2020 The druid Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Keyboard types.
+
+// This is a reasonable lint, but we keep signatures in sync with the
+// bitflags implementation of the inner Modifiers type.
+#![allow(clippy::trivially_copy_pass_by_ref)]
+
+use std::ops::{BitAnd, BitAndAssign, BitOr, BitOrAssign, BitXor, BitXorAssign, Not};
+
+pub use keyboard_types::{Code, KeyState, Location};
+
+pub type KbKey = keyboard_types::Key;
+
+/// Information about a keyboard event.
+///
+/// Note that this type is similar to [`KeyboardEvent`] in keyboard-types,
+/// but has a few small differences for convenience. It is missing the `state`
+/// field because that is already implicit in the event.
+///
+/// [`KeyboardEvent`]: keyboard_types::KeyboardEvent
+#[derive(Clone, Debug, Default, Eq, Hash, PartialEq)]
+pub struct KeyEvent {
+    /// Whether the key is pressed or released.
+    pub state: KeyState,
+    /// Logical key value.
+    pub key: KbKey,
+    /// Physical key position.
+    pub code: Code,
+    /// Location for keys with multiple instances on common keyboards.
+    pub location: Location,
+    /// Flags for pressed modifier keys.
+    pub modifiers: Modifiers,
+    /// True if the key is currently auto-repeated.
+    pub repeat: bool,
+    /// Events with this flag should be ignored in a text editor
+    /// and instead composition events should be used.
+    pub is_composing: bool,
+}
```

Comment:
in case you were curious we could save 8 bytes everywhere this is used by jamming `is_composing`, `repeat`, and `location` into spare bits on modifiers 🥴 🤷